A list of all open source software used in the analyses
-------------------------------------------------------


|Software   |Version         |Description                                  |URL                                           |
|:----------|:--------------:|:--------------------------------------------|:---------------------------------------------|
|Trimmomatic|v0.33/v0.36     |Cleaning and trimming of Illumina reads      |http://www.usadellab.org/cms/?page=trimmomatic|
|BWA        |v0.7            |Alignment of reads to reference sequence     |https://github.com/lh3/bwa                    |
|Picard     |v2.16           |Manipulation of high-throughput sequence data|https://broadinstitute.github.io/picard       |
|SAMtools   |v1.3/v1.9       |Manipulation of high-throughput sequence data|https://github.com/samtools/samtools          |
|BCFtools   |v1.4.1          |Manipulation of high-throughput sequence data|https://github.com/samtools/bcftools          |
|VCFtools   |v0.1.12b/v0.1.13|Manipulation of high-throughput sequence data|https://github.com/samtools/bcftools          |
|seqtk      |v1.2            |Manipulation of sequences in fasta and fastq format|https://github.com/lh3/seqtk|
|BBMap         |v37.47          |Merging overlapping reads into a single read	            |https://github.com/BioInfoTools/BBMap|
|SPAdes        |v3.9.1          |Genome assembly					    |https://github.com/ablab/spades|
|Mauve         |v2015-02-13     |Genome alignment					    |http://darlinglab.org/mauve/mauve.html|
|BEDtools      |v2.26.0         |A suite of command line tools for working with genome data |https://github.com/arq5x/bedtools2|
|RaxML         |v8.2.9/v8.2.12  |Analysis of phylogenies				    |https://github.com/stamatak/standard-RAxML|
|STRUCTURE     |v2.3.4          |Analysis of population genetic data using multilocus data  |https://web.stanford.edu/group/pritchardlab/structure.html|
|SplitsTree    |v4.14.5         |Computation of phylogenetic networks			    |http://www.splitstree.org|
|TreeCollapseCL|v4              |Manipulation of phylogenetic trees			    |http://emmahodcroft.com/TreeCollapseCL.html|
|ANGSD         |v0.924          |Analysis of next generation sequencing data		    |https://github.com/ANGSD/angsd|
|circos        |v0.69-6         |Circular visualization of the data			    |http://circos.ca|
|Plink         |v1.9            |Manipulation of high-thoughput sequence data		    |http://pngu.mgh.harvard.edu/purcell/plink|
|Blast2GO Basic|v5.2            |Functional annotation of genomic datasets		    |https://www.blast2go.com|
|EffectorP     |v1.0	        |Prediction of fungal effector proteins from secretomes	    |http://effectorp.csiro.au|
|SnIPRE        |downloaded 2019-01|Inference of selection in large gene datasets		    |https://bustamantelab.stanford.edu/lab-developed-software|
|D-Genies      |v1.2.0          |Display of dot plots of genome alignments		    |http://dgenies.toulouse.inra.fr|
|NUCmer        |v3              |Genome alignment					    |http://mummer.sourceforge.net|
|DIAMOND       |v0.9.22         |Local sequence alignment				    |https://github.com/bbuchfink/diamond|
|mstatspop     |v0.1	        |Analysis of population genomic data			    |https://github.com/CRAGENOMICA/mstatspop|
|Twisst        |downloaded 2018-02|Topology weighting					    |https://github.com/simonhmartin/twisst|



Languages and packages
----------------------
|Package|Version|Description|URL|
|:------|:-----:|:----------|:--|
|**python**|**v2.7/v3.6**||
|ETE|v3|Analysis and visualization of phylogenetic trees|http://etetoolkit.org|
|scikit-allel|v1.2.0|Manipulation and analysis of genomic data|https://scikit-allel.readthedocs.io/en/stable|
|**R**|**v3.4/v3.5/v3.6**||
|PopGenome|v2.6.1|Analysis of population genomic data|https://www.rdocumentation.org/packages/PopGenome/versions/2.7.1|

Other pipelines
---------------
|Description|DOI|
|:----------|:--------|
|Annotation of transposable elements|https://doi.org/10.1371/journal.pcbi.0010022|
|SpeedSeq: Detection of structural variants|https://doi.org/10.1038/nmeth.3505|

